crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs

```diff
@@ -0,0 +1,878 @@
+/// Utilities for sorting constant lists of string literals.
+///
+/// Examples where these are useful:
+/// - Sorting `__all__` in the global scope,
+/// - Sorting `__slots__` or `__match_args__` in a class scope
+use std::borrow::Cow;
+use std::cmp::Ordering;
+
+use ruff_python_ast as ast;
+use ruff_python_codegen::Stylist;
+use ruff_python_parser::{lexer, Mode, Tok, TokenKind};
+use ruff_python_stdlib::str::is_cased_uppercase;
+use ruff_python_trivia::leading_indentation;
+use ruff_source_file::Locator;
+use ruff_text_size::{Ranged, TextRange, TextSize};
+
+use is_macro;
+use itertools::Itertools;
+use natord;
+
+/// An enumeration of the different sorting styles
+/// currently supported for displays of string literals
+#[derive(Debug)]
+pub(super) enum SortingStyle {
+    Natural,
+    Isort,
+}
+
+impl SortingStyle {
+    pub(super) fn compare(&self, a: &str, b: &str) -> Ordering {
+        match self {
+            Self::Natural => natord::compare(a, b),
+            Self::Isort => IsortSortKey::from(a).cmp(&IsortSortKey::from(b)),
+        }
+    }
+}
+
+/// A struct to implement logic necessary to achieve
+/// an "isort-style sort".
+///
+/// An isort-style sort sorts items first according to their casing:
+/// `SCREAMING_SNAKE_CASE` names (conventionally used for global constants)
+/// come first, followed by CamelCase names (conventionally used for
+/// classes), followed by anything else. Within each category,
+/// a [natural sort](https://en.wikipedia.org/wiki/Natural_sort_order)
+/// is used to order the elements.
+struct IsortSortKey<'a> {
+    category: InferredMemberType,
+    value: &'a str,
+}
+
+impl Ord for IsortSortKey<'_> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.category
+            .cmp(&other.category)
+            .then_with(|| natord::compare(self.value, other.value))
+    }
+}
+
+impl PartialOrd for IsortSortKey<'_> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for IsortSortKey<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl Eq for IsortSortKey<'_> {}
```

Comment:
@BurntSushi - Do these `Ord` / `PartialOrd` / `PartialEq` / `Eq` implementations look right / consistent to you? (I've never properly learned the relationships between them.)

Reply:
(FWIW: these haven't changed in this PR, just moved to a different file, and Micha reviewed these in my previous PR)